onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -724,29 +425,29 @@ Status MatMulNBits::ComputeInternal(onnxruntime::webgpu::ComputeContext& context
   }
 
   // On FP32 only GPUs, integer math is faster than FP32 therefore always use DP4A independent of length of M.
-  if ((M >= kMinMForTileOptimization || y->DataType() == DataTypeImpl::GetType<float>() || nbits == 8 ||
+  if ((M >= kMinMForTileOptimization || y->DataType() == DataTypeImpl::GetType<float>() ||
        context.AdapterInfo().vendor == std::string_view{"qualcomm"}) &&
       CanApplyDP4AMatrixMatMulNBits(context, accuracy_level_, block_size, batch_count, N, K, components_a, has_zero_points)) {
     return ApplyDP4AMatrixMatMulNBits(a, b, scales, M, N, K, block_size, kMinMForTileOptimization, nbits, context, y);
   }
 
-  // TODO: Remvoe it once the 8bits is supported for the non-dp4 path.
-  ORT_ENFORCE(nbits == 4, "Only 4 bits are supported for the non-dp4 path for webgpu matmulnbits");
+  // zero_points has shape[N * CeilDiv(n_blocks_per_col * bits, 8)]. So here we need to check whether n_blocks_per_col is divisible by 8/nbits.
+  uint32_t zero_blocks_per_col = n_blocks_per_col % (8 / nbits) == 0 ? n_blocks_per_col : n_blocks_per_col + 1;
```

Comment:
the number of zero_points would be the same as n_blocks_per_col  always no ? its the same number as the number of scale items. If the data length is not divisble by block_size, the n_blocks_per_col should already contain 1 extra element.

Reply:
If zero_points has same type as A, it's not packed and has the same shape as Scales. But here, zero_points is stored as uint8_t as input B,  the shape becomes `[N * CeilDiv(n_blocks_per_col * bits, 8)]`. So we have to process that `n_blocks_per_col * bits` is not divisible by 8. It happens when bits is 4 and n_blocks_per_col is odd number. To make `mm_read_zero` works correctly, we make this change.